src/java.base/share/classes/java/lang/constant/ClassDesc.java

```diff
@@ -83,7 +83,7 @@ public sealed interface ClassDesc
      */
     static ClassDesc of(String name) {
         validateBinaryClassName(name);
-        return ClassDesc.ofDescriptor("L" + binaryToInternal(name) + ";");
+        return ClassDesc.ofDescriptor('L' + binaryToInternal(name) + ';');
```

Comment:
Do you think we should use the explicit unrolled StringBuilder, like:
```java
StringBuilder sb = new StringBuilder(name + 2);
sb.append('L').append(binaryToInternal(name)).append(';');
return ClassDesc.ofDescriptor(sb.toString());
```
Or we can use `ReferenceClassDescImpl.ofValidated` if we update `validateBinaryClassName` to reject leading, trailing, or consecutive `.`.

Reply:
Explicitly expanding the StringBuilder code will reduce the readability. I prefer a utility method that supports the three-parameter method concat(char, String, char)

Reply:
```java
final class StringConcatHelper {
    static String concat(char c0, String str, char c1) {
        if (str == null) {
            str = "null";
        }
        byte coder = (byte) (stringCoder(c0) | str.coder() | stringCoder(c1));
        int len = str.length() + 2;
        byte[] buf = newArray(len << coder);
        int last = buf.length - 1;
        if (coder == String.LATIN1) {
            buf[last] = (byte) (c1 & 0xFF);
        } else {
            StringUTF16.putChar(buf, last, c1);
        }
        prepend(last, coder, buf, c0, str);
        return new String(buf, coder);
    }
}

public interface JavaLangAccess {
    String concat(char c0, String str, char c1);
}


static ClassDesc ofInternalName(String name) {
    validateInternalClassName(name);
    return ClassDesc.ofDescriptor(JLA.concat('L', name, ';'));
}
```

This approach will be faster. Can we accept adding JLA methods to speed up startup?

Reply:
I think you can add this static method in somewhere in jdk.internal, and that place in jdk.internal calls JavaLangAccess; so users don't need to add their own JLA fields to call this method.

Reply:
Adding internal `concat` variants with `char` constants can be relatively noticeable for low-arity expressions, sure. How big is the win? And how much worse would a (more generally applicalbe) `concat(String, String, String)` fare? 

Doing the same trick for generated code might be a bit of a slippery slope and would cause an explosion of possible concat classes if we applied it generally. Supporting any permutation of interspersed constant of either `String` or `char` type would increase the upper bound of number of classes for a given arity `N` by a factor of `2^(N+1)`)). For just 1 arg concats that would increase the max number of generated classes from 7 to 28.

(I think these concat optimizations should be a separate PR)

Reply:
A separate PR has been created for ClassDesc related string concatenation optimization https://github.com/openjdk/jdk/pull/20705